File: scripts/serve_v3_compact.py
#!/usr/bin/env python3
"""Key-blind ONNX service for the compact V3 continuous-answer model."""

import base64
import io
import json
import os
import ssl
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlsplit

import numpy as np
import onnxruntime as ort
from PIL import Image

HEIGHT, WIDTH = 64, 192
FORBIDDEN = {"answerkey", "answer_key", "expected", "expectedanswer", "mathematicalanswer", "canonicaldigits"}


def contains_key_leak(value):
    if isinstance(value, dict):
        return any(str(key).replace("-", "").replace("_", "").lower() in FORBIDDEN or contains_key_leak(item) for key, item in value.items())
    if isinstance(value, list):
        return any(contains_key_leak(item) for item in value)
    return False


def decode_image(data_url):
    encoded = data_url.split(",", 1)[1] if "," in data_url else data_url
    raw = base64.b64decode(encoded, validate=True)
    if len(raw) > 3_000_000:
        raise ValueError("image exceeds 3 MB")
    image = Image.open(io.BytesIO(raw)).convert("L")
    scale = min(WIDTH / image.width, HEIGHT / image.height)
    resized = image.resize((max(1, round(image.width * scale)), max(1, round(image.height * scale))), Image.Resampling.BILINEAR)
    canvas = Image.new("L", (WIDTH, HEIGHT), 255)
    canvas.paste(resized, ((WIDTH - resized.width) // 2, (HEIGHT - resized.height) // 2))
    values = 1.0 - np.asarray(canvas, dtype=np.float32) / 255.0
    return values[None, None, :, :]


def softmax(values):
    values = values - np.max(values, axis=-1, keepdims=True)
    exp = np.exp(values)
    return exp / exp.sum(axis=-1, keepdims=True)


class Recognizer:
    def __init__(self, model_path):
        self.path = model_path
        self.session = ort.InferenceSession(model_path, providers=["CPUExecutionProvider"])
        self.input_name = self.session.get_inputs()[0].name
        self.lock = threading.Lock()

    def run(self, items):
        batch = np.concatenate([decode_image(item["continuousImageDataUrl"]) for item in items], axis=0)
        with self.lock:
            length_logits, tens_logits, ones_logits = self.session.run(None, {self.input_name: batch})
        length_probs, tens_probs, ones_probs = map(softmax, (length_logits, tens_logits, ones_logits))
        results = []
        for index, item in enumerate(items):
            length = int(length_probs[index].argmax()) + 1
            ones = int(ones_probs[index].argmax())
            probabilities = [float(length_probs[index, length - 1]), float(ones_probs[index, ones])]
            if length == 2:
                tens = int(tens_probs[index].argmax())
                read = f"{tens}{ones}"
                probabilities.append(float(tens_probs[index, tens]))
            else:
                read = str(ones)
            candidates = []
            for digit in range(10):
                components = [float(length_probs[index, 0]), float(ones_probs[index, digit])]
                candidates.append({
                    "read": str(digit),
                    "jointProbability": components[0] * components[1],
                    "minComponentProbability": min(components),
                })
            for left in range(10):
                for right in range(10):
                    components = [
                        float(length_probs[index, 1]),
                        float(tens_probs[index, left]),
                        float(ones_probs[index, right]),
                    ]
                    candidates.append({
                        "read": f"{left}{right}",
                        "jointProbability": components[0] * components[1] * components[2],
                        "minComponentProbability": min(components),
                    })
            candidates.sort(key=lambda row: (row["jointProbability"], row["minComponentProbability"]), reverse=True)
            results.append({
                "id": item.get("id"), "questionNum": item.get("questionNum"), "read": read,
                "frameIndex": item.get("frameIndex"),
                "meanComponentProbability": sum(probabilities) / len(probabilities),
                "minComponentProbability": min(probabilities),
                "topCandidates": [{
                    **candidate,
                    "jointProbability": round(candidate["jointProbability"], 8),
                    "minComponentProbability": round(candidate["minComponentProbability"], 6),
                } for candidate in candidates[:5]],
                "model": "scangrade-v3-compact-continuous", "keyBlind": True,
            })
        return results


MODEL = os.environ.get("SCANGRADE_V3_COMPACT_MODEL", "private-evidence/models/v3-sequence-live/model.onnx")
RECOGNIZER = Recognizer(MODEL)
ALLOWED = {item.strip() for item in os.environ.get("SCANGRADE_V3_ALLOWED_ORIGINS", "").split(",") if item.strip()}
TOKEN = os.environ.get("SCANGRADE_V3_TOKEN", "")


class Handler(BaseHTTPRequestHandler):
    def origin_allowed(self):
        origin = self.headers.get("Origin", "")
        return not origin or not ALLOWED or origin in ALLOWED

    def authorized(self):
        return not TOKEN or self.headers.get("Authorization", "") == f"Bearer {TOKEN}"

    def cors(self):
        origin = self.headers.get("Origin", "")
        if origin and (not ALLOWED or origin in ALLOWED):
            self.send_header("Access-Control-Allow-Origin", origin)
            self.send_header("Vary", "Origin")

    def response(self, status, payload):
        data = json.dumps(payload).encode()
        self.send_response(status)
        self.cors()
        self.send_header("Content-Type", "application/json")
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_OPTIONS(self):
        if not self.origin_allowed():
            return self.response(403, {"ok": False, "error": "origin not allowed"})
        self.send_response(204)
        self.cors()
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
        self.end_headers()

    def do_GET(self):
        if urlsplit(self.path).path == "/health":
            self.response(200, {"ok": True, "model": "scangrade-v3-compact-continuous", "modelPath": MODEL})
        else:
            self.response(404, {"ok": False, "error": "not found"})

    def do_POST(self):
        if urlsplit(self.path).path != "/v3/recognize":
            return self.response(404, {"ok": False, "error": "not found"})
        if not self.origin_allowed():
            return self.response(403, {"ok": False, "error": "origin not allowed"})
        if not self.authorized():
            return self.response(401, {"ok": False, "error": "unauthorized"})
        try:
            length = int(self.headers.get("Content-Length", "0"))
            if length <= 0 or length > 30_000_000:
                return self.response(413, {"ok": False, "error": "invalid request size"})
            payload = json.loads(self.rfile.read(length))
            if contains_key_leak(payload):
                return self.response(400, {"ok": False, "error": "answer-key fields are forbidden"})
            items = payload.get("items", [])
            if not isinstance(items, list) or not 1 <= len(items) <= 64:
                return self.response(400, {"ok": False, "error": "items must contain 1-64 answers"})
            if any(not item.get("continuousImageDataUrl") for item in items):
                return self.response(400, {"ok": False, "error": "continuousImageDataUrl is required"})
            self.response(200, {"ok": True, "results": RECOGNIZER.run(items)})
        except Exception as error:
            self.response(500, {"ok": False, "error": str(error)})

    def log_message(self, format, *args):
        print(f"[v3-compact] {self.address_string()} {format % args}")


if __name__ == "__main__":
    host = os.environ.get("SCANGRADE_V3_HOST", "127.0.0.1")
    port = int(os.environ.get("PORT", os.environ.get("SCANGRADE_V3_PORT", "8767")))
    server = ThreadingHTTPServer((host, port), Handler)
    tls_cert = os.environ.get("SCANGRADE_V3_TLS_CERT", "")
    tls_key = os.environ.get("SCANGRADE_V3_TLS_KEY", tls_cert)
    protocol = "http"
    if tls_cert:
        context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        context.load_cert_chain(tls_cert, tls_key)
        server.socket = context.wrap_socket(server.socket, server_side=True)
        protocol = "https"
    print(f"V3 compact recognizer listening on {protocol}://{host}:{port}; model={MODEL}")
    server.serve_forever()

File: scripts/serve_v3_compact.mjs
#!/usr/bin/env node
import fs from 'node:fs/promises'
import http from 'node:http'
import https from 'node:https'
import path from 'node:path'
import { createCanvas, loadImage } from 'canvas'
import * as ort from 'onnxruntime-web'

const root = path.resolve(import.meta.dirname, '..')
const modelPath = path.resolve(root, process.env.SCANGRADE_V3_COMPACT_MODEL || 'private-evidence/models/v3-sequence-live/model.onnx')
const host = process.env.SCANGRADE_V3_HOST || '127.0.0.1'
const port = Number(process.env.SCANGRADE_V3_PORT || 8767)
const tlsCertPath = process.env.SCANGRADE_V3_TLS_CERT || ''
const tlsKeyPath = process.env.SCANGRADE_V3_TLS_KEY || tlsCertPath
const allowedOrigins = new Set(String(process.env.SCANGRADE_V3_ALLOWED_ORIGINS || '').split(',').map((item) => item.trim()).filter(Boolean))
const requiredToken = String(process.env.SCANGRADE_V3_TOKEN || '')
const forbidden = new Set(['answerkey', 'expected', 'expectedanswer', 'mathematicalanswer', 'canonicaldigits'])
ort.env.wasm.numThreads = 1
ort.env.wasm.simd = false
const session = await ort.InferenceSession.create(await fs.readFile(modelPath), { executionProviders: ['wasm'] })

function hasKeyLeak(value) {
  if (Array.isArray(value)) return value.some(hasKeyLeak)
  if (!value || typeof value !== 'object') return false
  return Object.entries(value).some(([key, item]) => forbidden.has(key.replace(/[-_]/g, '').toLowerCase()) || hasKeyLeak(item))
}

async function tensorFor(dataUrl) {
  const image = await loadImage(dataUrl)
  const scale = Math.min(192 / image.width, 64 / image.height)
  const width = Math.max(1, Math.round(image.width * scale))
  const height = Math.max(1, Math.round(image.height * scale))
  const canvas = createCanvas(192, 64)
  const context = canvas.getContext('2d')
  context.fillStyle = '#fff'
  context.fillRect(0, 0, 192, 64)
  context.drawImage(image, Math.floor((192 - width) / 2), Math.floor((64 - height) / 2), width, height)
  const rgba = context.getImageData(0, 0, 192, 64).data
  const values = new Float32Array(192 * 64)
  for (let index = 0; index < values.length; index += 1) {
    const offset = index * 4
    const gray = .299 * rgba[offset] + .587 * rgba[offset + 1] + .114 * rgba[offset + 2]
    values[index] = 1 - gray / 255
  }
  return values
}

function softmax(values) {
  const max = Math.max(...values)
  const exp = values.map((value) => Math.exp(value - max))
  const total = exp.reduce((sum, value) => sum + value, 0)
  return exp.map((value) => value / total)
}

function argmax(values) {
  let best = 0
  for (let index = 1; index < values.length; index += 1) if (values[index] > values[best]) best = index
  return best
}

async function recognize(items) {
  const batch = new Float32Array(items.length * 192 * 64)
  for (let index = 0; index < items.length; index += 1) batch.set(await tensorFor(items[index].continuousImageDataUrl), index * 192 * 64)
  const output = await session.run({ image: new ort.Tensor('float32', batch, [items.length, 1, 64, 192]) })
  return items.map((item, index) => {
    const length = softmax(Array.from(output.length.data.slice(index * 2, index * 2 + 2)))
    const tens = softmax(Array.from(output.tens.data.slice(index * 10, index * 10 + 10)))
    const ones = softmax(Array.from(output.ones.data.slice(index * 10, index * 10 + 10)))
    const lengthIndex = argmax(length)
    const tensIndex = argmax(tens)
    const onesIndex = argmax(ones)
    const probabilities = lengthIndex === 1 ? [length[1], tens[tensIndex], ones[onesIndex]] : [length[0], ones[onesIndex]]
    const candidates = []
    for (let digit = 0; digit < 10; digit += 1) {
      const components = [length[0], ones[digit]]
      candidates.push({
        read: String(digit),
        jointProbability: components[0] * components[1],
        minComponentProbability: Math.min(...components),
      })
    }
    for (let left = 0; left < 10; left += 1) {
      for (let right = 0; right < 10; right += 1) {
        const components = [length[1], tens[left], ones[right]]
        candidates.push({
          read: `${left}${right}`,
          jointProbability: components[0] * components[1] * components[2],
          minComponentProbability: Math.min(...components),
        })
      }
    }
    candidates.sort((a, b) => b.jointProbability - a.jointProbability || b.minComponentProbability - a.minComponentProbability)
    return {
      id: item.id, questionNum: item.questionNum, frameIndex: item.frameIndex ?? null,
      read: lengthIndex === 1 ? `${tensIndex}${onesIndex}` : String(onesIndex),
      meanComponentProbability: probabilities.reduce((sum, value) => sum + value, 0) / probabilities.length,
      minComponentProbability: Math.min(...probabilities),
      topCandidates: candidates.slice(0, 5).map((candidate) => ({
        ...candidate,
        jointProbability: Number(candidate.jointProbability.toFixed(8)),
        minComponentProbability: Number(candidate.minComponentProbability.toFixed(6)),
      })),
      model: 'scangrade-v3-compact-continuous', keyBlind: true,
    }
  })
}

function cors(response, request) {
  const origin = request.headers.origin
  if (origin && (!allowedOrigins.size || allowedOrigins.has(origin))) {
    response.setHeader('Access-Control-Allow-Origin', origin)
    response.setHeader('Vary', 'Origin')
  }
}

function originAllowed(request) {
  const origin = request.headers.origin
  return !origin || !allowedOrigins.size || allowedOrigins.has(origin)
}

function authorized(request) {
  if (!requiredToken) return true
  return request.headers.authorization === `Bearer ${requiredToken}`
}

function reply(response, request, status, value) {
  cors(response, request)
  response.writeHead(status, {
    'Content-Type': 'application/json',
    'Cache-Control': 'no-store',
    'X-Content-Type-Options': 'nosniff',
  })
  response.end(JSON.stringify(value))
}

const requestHandler = async (request, response) => {
  if (request.method === 'OPTIONS') {
    if (!originAllowed(request)) return reply(response, request, 403, { ok: false, error: 'origin not allowed' })
    cors(response, request)
    response.writeHead(204, { 'Access-Control-Allow-Headers': 'Content-Type, Authorization', 'Access-Control-Allow-Methods': 'GET,POST,OPTIONS' })
    return response.end()
  }
  if (request.method === 'GET' && request.url === '/health') return reply(response, request, 200, { ok: true, model: 'scangrade-v3-compact-continuous', modelPath })
  if (request.method !== 'POST' || request.url !== '/v3/recognize') return reply(response, request, 404, { ok: false, error: 'not found' })
  if (!originAllowed(request)) return reply(response, request, 403, { ok: false, error: 'origin not allowed' })
  if (!authorized(request)) return reply(response, request, 401, { ok: false, error: 'unauthorized' })
  let size = 0
  const chunks = []
  for await (const chunk of request) {
    size += chunk.length
    if (size > 30_000_000) return reply(response, request, 413, { ok: false, error: 'request too large' })
    chunks.push(chunk)
  }
  try {
    const payload = JSON.parse(Buffer.concat(chunks).toString('utf8'))
    if (hasKeyLeak(payload)) return reply(response, request, 400, { ok: false, error: 'answer-key fields are forbidden' })
    if (!Array.isArray(payload.items) || payload.items.length < 1 || payload.items.length > 64 || payload.items.some((item) => !item.continuousImageDataUrl)) {
      return reply(response, request, 400, { ok: false, error: 'items must contain 1-64 continuous images' })
    }
    return reply(response, request, 200, { ok: true, results: await recognize(payload.items) })
  } catch (error) {
    return reply(response, request, 500, { ok: false, error: String(error?.message || error) })
  }
}
const server = tlsCertPath
  ? https.createServer({ cert: await fs.readFile(tlsCertPath), key: await fs.readFile(tlsKeyPath) }, requestHandler)
  : http.createServer(requestHandler)
const protocol = tlsCertPath ? 'https' : 'http'
server.listen(port, host, () => console.log(`V3 compact recognizer listening on ${protocol}://${host}:${port}; model=${modelPath}`))
